### ISSUE TITLE
Rule proposal: `no-iterable-to-array`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1403]
Got the idea implementing #1401, it's kind of like what [`no-useless-spread`](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1395) do, so maybe this may add to that one, but let's discuss first.

1. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#built-in_apis_accepting_iterables), these builtins accepts iterable should use iterable directly.
2. Objects in `for..of` should not transform to array.
3. When using methods from array, if iterable already have it, should not transform to array, see [Iterator Helpers proposal](https://github.com/tc39/proposal-iterator-helpers).

## Fail

```js
new Map([...iterable])
new WeakMap([...iterable])
new Set([...iterable])
new WeakSet([...iterable])
Promise.all([...iterable])
Promise.race([...iterable])
Promise.allSettled([...iterable]) // Missing on MDN
Array.from([...iterable]) // I think no one does this though
```

```js
for (const foo of [...iterable]);
```

```js
[...document.querySelectorAll('*')].forEach(() => {})
```

## Pass

```js
new Map(iterable)
new WeakMap(iterable)
new Set(iterable)
new WeakSet(iterable)
Promise.all(iterable)
Promise.race(iterable)
Promise.allSettled(iterable)
Array.from(iterable)
```

```js
for (const foo of iterable);
```

```js
document.querySelectorAll('*').forEach(() => {})
```

**Comment by sindresorhus** [comment]
Sounds good. I think we can just add it to `no-useless-spread` as this is also about preventing useless use of spread.

**Comment by fregante** [comment]
This is slightly different tough:

```js
[...document.querySelectorAll('*')].forEach(() => {})
```

Indeed here it's useless, but `.forEach` is the only available method so converting to an array is required for others:

```js
[...document.querySelectorAll('*')].map(() => {})
```
